docs/blog/posts/write-code-to-refactor.md

```diff
@@ -0,0 +1,402 @@
+    writer.writeheader()
+    writer.writerows(data)
+```
+
+But, we could also use pandas. 
+
+```python
+import pandas as pd
+
+data = [{"key1": 2, "key4": 10}, {"key3": 3}, {"key4": 4}]
+df = pd.DataFrame(data)
+df.to_csv("output.csv", index=False)
+```
+
+This is a toy example here so the differences between the two code examples are minimal. But we can extrapolate this to more complicated procedures - for instance a model training job. By using a standardized library, we're able to prevent a lot of unforseen issues and take advantage of the time and effort people have spent to make their code robust and reliable. Furthermore, a standardized library is often frequently shared amongst different projects, making it easy to scale up the complexity of the codebase down the line.
+
+### Manage Complexity
+
+Let's give a small example - let's say we'd like to write a class that's able to support different providers (eg. Hugging Face, OpenAI ) and run an embedding job on them. An initial draft might look like this
+
+```python
+class EmbeddingModel():
+  def __init__(self,model_name,model_type):
+    if model_type == "HuggingFace":
+      self.model_type = "HuggingFace"
+      self.model = SentenceTransformer(model_name)
+    elif model_type == "OpenAI":
+      self.model_type = "OpenAI"
+      self.model = OpenAI()
+    else:
+      raise ValueError(f"Invalid Model Type of {model_type} was provided. Please make sure to only pass Hugging Face or OpenAI as valid values")
+```
+
+This class is **very prone to errors** because we're relying on a single string to determine the logic. Imagine how much more complex we'd need to make the control flow when we add more options and complex configurations. 
+
+In order to prevent these complex errors from happening, we can use a `@classmethod` and a `enum` in order to prevent this from happening. A  `classmethod` allows us to instantiate a class with the right configuration so that we don't run into these complex edge cases. Furthermore, by pre-defining the values that a user can pass in, we can also support more complex initialisation logic.
+
+```python
+import enum
+
+class Provider(enum.Enum):
+    HUGGINGFACE = "HuggingFace"
+    OPENAI = "OpenAI"
+    COHERE = "Cohere"
+
+class EmbeddingModel:
+    def __init__(self,model_name:str,model_type=Provider):
+        self.model_name=model_name
+        self.model_type =model_type
+	
+    @classmethod
+    def from_openai(cls, model_name: str):
+        return cls(model_name, model_type=Provider.OPENAI)
+    
+    @classmethod
+    def from_hf(cls, model_name: str):
+        return cls(
+            model_name,
+            provider=Provider.HUGGINGFACE,
+        )
+```
+
+Small things like class methods and enums go a long way in ensuring that code that we write is safe and works as intended.
+
+### Write Simple Functions
+
+The functions I write have ideally three characteristics
+
+- They do one thing and their name reflects it
```

Comment:
function names are verbs or actions, get_x, compute_y
and types are nowns "results"

vs `def results(user_id)`